### ISSUE TITLE
Debugger ignores some try ... except ... statements

### ISSUE THREAD

**Issue by ARF1** [Microsoft/PTVS#17]
Using "Start without debugging" the following code executes without problems, but when running with debugging the program breaks with a "No module named invalidImport" error. It presents the options "Break" or "Continue" as for all non-recoverable errors. If I choose "Continue" the program continues without any issue.

The debugger somehow seems to ignore the `try ...  except ...` statement in the context manager while python correctly catches the error.

Clearly PTVS debugging thus should not break here.

```
from contextlib import contextmanager

@contextmanager
def ignoring(*exceptions):
    try:
        yield
    except exceptions:
        pass

with ignoring(ImportError):
    import invalidImport

print "Test"
```

**Comment by int19h** [comment]
The default is to break when exception is unhandled, but determining what exactly "unhandled" means is tricky. Python doesn't have two-pass exception handling that is common to most other languages, and so there's no stage at which the stack hasn't been unwound yet, but the destination of unwinding is already known. So the only way to figure out whether the exception is going to be handled at the point where it's being thrown is to try to simulate the unwinding.

Currently, we do this by looking at the source code on AST level, matching it against current line numbers on the call stack (you can see this in GetHandledExceptionRanges [here](../blob/master/Python/Product/Debugger/Debugger/PythonProcess.cs#L504)). This only handles direct name-based lookup, so code like yours, which flows the exception list through a variable, doesn't work right.

Making it work for everything would be pretty complicated. On source level, we could try to rely on our code analysis engine to figure out the types instead of just doing AST parsing, but that won't cover all cases reliably, and we don't necessarily have analysis for any file that might have an `except` block in it. Or we could eval the expressions in `except` at runtime, but they potentially have side effects, and there's no reliable way to tell whether they do or not.

**Comment by mrocklin** [comment]
Note that `ignoring` (or rather `ignored`) has been added to the stdlib.  So you may start seeing this problem more frequently in the future.

https://hg.python.org/cpython/rev/406b47c64480

**Comment by zooba** [comment]
We should enhance our support here to handle at least the `ignored` context manager.

Handling arbitrary context managers is probably too complicated unless we emit an exception map from analysis before starting debugging.

**Comment by ARF1** [comment]
@zooba I was wondering whether there is any movement on this. This feature been in the standard library since python 3.4 as [`contextlib.suppress`](https://docs.python.org/3.6/library/contextlib.html#contextlib.suppress). It wound be really great if at least this could be special-cased, so that we all can start writing cleaner code. At present debugging code using `suppress` is next to impossible. It seems weird your PR has been available for quite some time but never made into the code base.

Is there anything users can do as a stop-gap measure?

@mrocklin Do you know when the name changed from `ignored` to `suppress`? Did the `ignored` version ever get shipped?

**Comment by mrocklin** [comment]
No clue.  It seems more likely that I confused the name.

**Comment by ARF1** [comment]
@mrocklin No, you were quite right. Upon futher investigation I found that this feature was proposed and implemented in [Python Issue15806](http://bugs.python.org/issue15806) but was never shipped. It was renamed in [Issue19266](http://bugs.python.org/issue19266) from `ignored` to `suppress` and shipped in this form with python 3.4.

**Comment by huguesv** [comment]
We have a new debug engine in the 15.8 release, if you still see this problem, please open an issue at https://github.com/Microsoft/ptvsd
